gdal/frmts/fits/fitsdataset.cpp

```diff
@@ -599,6 +734,689 @@ GDALDataset *FITSDataset::Create( const char* pszFilename,
+/*                        DeleteNoDataValue()                           */
+/************************************************************************/
+
+CPLErr FITSRasterBand::DeleteNoDataValue()
+
+{
+    poDS->LoadGeoreferencingAndPamIfNeeded();
+
+    if( !poDS->bNoDataSet )
+        return CE_None;
+
+    poDS->bNoDataSet = false;
+    poDS->dfNoDataValue = -9999.0;
+
+    poDS->bNoDataChanged = true;
+
+    bNoDataSet = false;
+    dfNoDataValue = -9999.0;
+    return CE_None;
+}
+
+/************************************************************************/
+/*                     LoadGeoreferencingAndPamIfNeeded()               */
+/************************************************************************/
+
+void FITSDataset::LoadGeoreferencingAndPamIfNeeded()
+
+{
+    int status = 0;
+    int bitpix;
+    double dfScale, dfOffset;
+    double crpix1, crpix2, crval1, crval2, cdelt1, cdelt2, pc[4], cd[4];
+    double aRadius, cRadius, invFlattening = 0.0;
+    char target[16], ctype[9];
+    std::string pszGeogName, pszDatumName, projName;
+
+    const double PI = std::atan(1.0)*4;
+    const double DEG2RAD = PI / 180.;
+
+    if( !m_bReadGeoTransform && !m_bLoadPam )
+        return;
+
+/* -------------------------------------------------------------------- */
+/*      Get the transform from the FITS file.                           */
+/* -------------------------------------------------------------------- */
+    if( m_bReadGeoTransform )
+    {
+        m_bReadGeoTransform = false;
+
+        fits_read_key(hFITS, TSTRING, "OBJECT", target, nullptr, &status);
+        if( status )
+        {
+            strncpy(target, "Undefined", 10);
+            CPLError(CE_Failure, CPLE_AppDefined,
+                 "OBJECT keyword is missing");
+            status = 0;
+        }
+
+        pszGeogName.assign("GCS_");
+        pszGeogName.append(target);
+        pszDatumName.assign("D_");
+        pszDatumName.append(target);
+
+        fits_read_key(hFITS, TDOUBLE, "A_RADIUS", &aRadius, nullptr, &status);
+        if( status )
+        {
+            CPLError(CE_Failure, CPLE_AppDefined,
+                 "No Radii keyword available, metadata will not contain DATUM information.");
+            status = 0;
+        } else {
+            fits_read_key(hFITS, TDOUBLE, "C_RADIUS", &cRadius, nullptr, &status);
+            if( aRadius != cRadius )
+            {
+                invFlattening = aRadius / ( aRadius - cRadius );
+            }
+        }
+
+        /* Waiting for linear keywords standardization only deg ctype are used */
+        /* Check if WCS are there */
+        fits_read_key(hFITS, TSTRING, "CTYPE1", ctype, nullptr, &status);
+        if ( !status ) {
+            /* Check if angular WCS are there */
+            if ( strstr(ctype, "LN") )
+            {
+                /* Reading reference points */
+                fits_read_key(hFITS, TDOUBLE, "CRPIX1", &crpix1, nullptr, &status);
```

Comment:
perhaps all those variables set by fits_read_key should be zero initialized during their declaration if they don't exist and fits_read_key() doesn't zero terminate them

Reply:
I have added one more check of status on CRPIX/CRVAL keywords: again, if they are not there or the read function throws an error the georef is skipped (https://github.com/OSGeo/gdal/pull/1298/commits/c22caadc8e8df76430f53983d63bb9cb3fd0a517).